datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -563,10 +565,10 @@ fn push_down_join(
 /// * `on_filters` filters from the join ON clause that have not already been
 ///   identified as join predicates
 ///
-fn infer_join_predicates(
+fn infer_join_predicates<E: AsRef<Expr>>(
     join: &Join,
-    predicates: &[Expr],
-    on_filters: &[Expr],
+    predicates: &[E],
```

Comment:
This can also be written as `impl IntoIterator<Item = &'a Expr>`
but this is much less flexible. It allows only iteration and only once. Slice can be passed to multiple such functions, iterated over or random-accessed.